File: backend/services/symptom_log_service.py
from typing import Optional
from uuid import UUID

from sqlalchemy.orm import Session

from repositories.symptom_log_repository import SymptomLogRepository
from schemas.symptom_log import SymptomLogCreate, SymptomLogResponse, SymptomLogUpdate


class SymptomLogService:
    """Business logic for symptom logs."""

    def __init__(self):
        self.repo = SymptomLogRepository()

    def get_symptom_log(self, db: Session, log_id: UUID) -> Optional[SymptomLogResponse]:
        log = self.repo.get_by_id(db, log_id)
        if not log:
            return None
        return SymptomLogResponse.model_validate(log)

    def create_symptom_log(self, db: Session, data: SymptomLogCreate) -> SymptomLogResponse:
        log = self.repo.create(db, data)
        return SymptomLogResponse.model_validate(log)

    def get_symptom_logs_by_username(self, db: Session, username: str) -> list[SymptomLogResponse]:
        logs = self.repo.get_by_username(db, username)
        return [SymptomLogResponse.model_validate(log) for log in logs]

    def update_symptom_log(self, db: Session, log_id: UUID, data: SymptomLogUpdate) -> Optional[SymptomLogResponse]:
        updated = self.repo.update(db, log_id, data)
        if not updated:
            return None
        return SymptomLogResponse.model_validate(updated)

    def delete_symptom_log(self, db: Session, log_id: UUID) -> bool:
        return self.repo.delete(db, log_id)

File: frontend/src/api/symptom_log_service.ts
import { apiClient, ApiError } from './client';

export interface CreateSymptomLogPayload {
  symptom_id: string; 
  intensity: number; 
  timestamp: string; 
  duration?: number; 
  notes?: string;
  username: string;
}

export interface UpdateSymptomLogPayload {
  intensity?: number;
  timestamp?: string;
  duration?: number;
  notes?: string;
}

export interface SymptomLogResponse {
  id: string;
  symptom_id: string; 
  intensity: number;
  timestamp: string;
  duration: number | null;
  notes: string | null;
  username: string;
  created_at: string;
}

export const symptomLogService = {

  // create_symptom_log()
  async createSymptomLog(payload: CreateSymptomLogPayload): Promise<SymptomLogResponse> {
    try {
      const { data } = await apiClient.post<SymptomLogResponse>('/symptom-logs/', payload);
      return data;
    } catch (err: any) {
      throw new ApiError(err.response?.data?.detail ?? 'Failed to create symptom log');
    }
  },

  // get_my_symptom_logs()
  async getMySymptomLogs(): Promise<SymptomLogResponse[]> {
    try {
      const { data } = await apiClient.get<SymptomLogResponse[]>('/symptom-logs/user/me');
      return data;
    } catch (err: any) {
      throw new ApiError(err.response?.data?.detail ?? 'Failed to fetch symptom logs');
    }
  },

  // update_symptom_log()
  async updateSymptomLog(logId: string, payload: UpdateSymptomLogPayload): Promise<SymptomLogResponse> {
    try {
      const { data } = await apiClient.put<SymptomLogResponse>(`/symptom-logs/${logId}`, payload);
      return data;
    } catch (err: any) {
      throw new ApiError(err.response?.data?.detail ?? `Failed to update symptom log ${logId}`);
    }
  },

  // get_symptom_log()
  async getSymptomLogById(logId: string): Promise<SymptomLogResponse> {
    try {
      const { data } = await apiClient.get<SymptomLogResponse>(`/symptom-logs/${logId}`);
      return data;
    } catch (err: any) {
      throw new ApiError(err.response?.data?.detail ?? `Failed to fetch symptom log ${logId}`);
    }
  },

  // delete_symptom_log()
  async deleteSymptomLog(logId: string): Promise<void> {
    try {
      await apiClient.delete(`/symptom-logs/${logId}`);
    } catch (err: any) {
      throw new ApiError(err.response?.data?.detail ?? `Failed to delete symptom log ${logId}`);
    }
  },
};